src/libraries/Microsoft.PowerFx.Repl/Modules/ModuleLoadContext.cs

```diff
@@ -0,0 +1,238 @@
+using Microsoft.PowerFx.Core.Glue;
+using Microsoft.PowerFx.Core.Types;
+using Microsoft.PowerFx.Syntax;
+using YamlDotNet.Serialization;
+
+namespace Microsoft.PowerFx.Repl
+{
+    internal static class ModuleLoadContextExtentions
+    {   
+        // Load a module from disk. 
+        // This may throw on hard error (missing file), or return null with compile errors. 
+        public static async Task<Module> LoadFromFileAsync(this ModuleLoadContext context, string path, List<ExpressionError> errors)
+        {
+            var dir = Path.GetDirectoryName(path);
+            var name = Path.GetFileName(path);
+
+            var loader = new FileLoader(dir);
+
+            return await context.LoadAsync(name, loader, errors).ConfigureAwait(false);
+        }
+    }
+
+    /// <summary>
+    /// Context for loading a single module. 
+    /// This can ensure the module dependencies don't have cycles. 
+    /// </summary>
+    internal class ModuleLoadContext
+    {
+        // Core symbols we resolve against. 
+        // This is needed for all the builtins. 
+        private readonly ReadOnlySymbolTable _commonIncomingSymbols;
+
+        // Set of modules we visited - used to detect cycles. 
+        // String is the module's full path.
+        private readonly HashSet<string> _visited = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+
+        public ModuleLoadContext(ReadOnlySymbolTable commonIncomingSymbols)
+        {
+            _commonIncomingSymbols = commonIncomingSymbols;
+        }
+
+        /// <summary>
+        /// Load and return a resolved module.  May throw on errors. 
+        /// </summary>
+        /// <param name="name">name to resolve via loader parameter.</param>
+        /// <param name="loader">file loader to resolve the name and any further imports.</param>
+        /// <param name="errors">error collection to populate if there are errors in parsing, binding, etc. </param>
+        /// <returns>A module. Or possible null if there are errors. Or may throw. </returns>        
+        public async Task<Module> LoadAsync(string name, IFileLoader loader, List<ExpressionError> errors)
+        {
+            (var poco, var loader2) = await loader.LoadAsync(name).ConfigureAwait(false);
+
+            string fullPathIdentity = poco.Src_Filename.ToLowerInvariant();
+            if (!_visited.Add(fullPathIdentity))
+            {
+                // Already loaded
+                throw new InvalidOperationException($"Circular reference: {name}");
+            }
+
+            // Load all imports first. 
+
+            var symbolList = new List<ReadOnlySymbolTable>
+            {
+                _commonIncomingSymbols
+            };
+            var unique = new ConflictTracker();
+
+            if (poco.Imports != null)
+            {
+                foreach (var import in poco.Imports)
+                {
+                    int count = 0;
+                    string file = import.File;
```

Comment:
>import.File [](http://example.com/codeflow?start=34&length=11)

Not for v1, but we probably want caching for e.g. diamond imports

Reply:
added test for diamond. And doing some caching.